wave_optimization/include/wave/optimization/factor_graph/variable.hpp

```diff
@@ -0,0 +1,56 @@
+#ifndef WAVE_OPTIMIZATION_FACTOR_GRAPH_VARIABLE_HPP
+#define WAVE_OPTIMIZATION_FACTOR_GRAPH_VARIABLE_HPP
+
+#include "wave/utils/utils.hpp"
+
+namespace wave {
+
+struct Variable {
+    size_t id;
+    size_t size;
+    VecX data;
+
+    Variable() : id{0}, size{0}, data{0} {}
+    Variable(size_t id, size_t size) : id{id}, size{size}, data{size} {}
+
+    std::string toString() {
+        std::ostringstream oss;
+
+        oss << "[";
+        oss << "id: " << this->id << ", ";
+        oss << "size: " << this->size;
+        oss << "]";
+
+        return oss.str();
+    }
+
+    void print() {
+        std::cout << this->toString() << std::endl;
+    }
+
+    bool operator<(const Variable &other) const {
+        if (this->id < other.id) {
+            return true;
+        }
+        return false;
+    }
+
+    VecX &getData() {
+        return data;
+    }
+};
+
+struct PoseVar : Variable {
+    PoseVar(size_t id) : Variable{id, 6} {
```

Comment:
I think I see what you mean.  Yes, if you have a case like this
```cpp
struct Base {};
struct A : public Base {
    ChrisType data;
}
struct DerivedB : public Base {
    LeoType data;
}
std::vector<Base*> v;
v.push_back(new DerivedA());
v.push_back(new DerivedB());
```
You will not be able to directly access `data` through the `Base*`. But `Base` only needs to provide an _interface_, not expose `data` directly. See the chapter on inheritance in _Effective C++_.

There are a variety of possible solutions involving virtual functions, function pointers, and/or templates. 

You never have to "store the variable type in the class". I agree that would be silly.